Implement new plot directive for clustering visualizers [Issue #687]

Kautumn06 opened the pull request:
Adds the new plot directive from PR #446 in the documentation for the `KElbowVisualizer`, `SilhouetteVisualizer`, and `InterclusterDistance`. In addition, I deleted the scripts originally used to generate the images in the docs, along with the images.

I'll add comments to help explain the changes. However, please just let me know if anyone has any questions!

Quick Update: the tests originally failed which I think may have been because I initially set `k` in the `InterclusterDistance` visualizer, similar to how `k` is set in `KElbowVisualizer`. 
```
model = KMeans()
visualizer = InterclusterDistance(model, k=9)
```

And while it did create the image when I ran `make html`, the tests failed. I think this may be because `k` is not actually a defined parameter for the `InterclusterDistance` visualizer, which is why I pushed a new commit with the following update:
```
model = KMeans(9)
visualizer = InterclusterDistance(model)
```

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -8,21 +8,23 @@ The ``KElbowVisualizer`` implements the "elbow" method to help data scientists s
 To demonstrate, in the following example the ``KElbowVisualizer`` fits the ``KMeans`` model for a range of :math:`K` values from 4 to 11 on a sample two-dimensional dataset with 8 random clusters of points. When the model is fit with 8 clusters, we can see an "elbow" in the graph, which in this case we know to be the optimal number.
 
 .. plot::
+    :context: close-figs
```
Added the recommended `:context:` and `alt` settings to the plot directive.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -8,21 +8,23 @@ The ``KElbowVisualizer`` implements the "elbow" method to help data scientists s
-    # Create synthetic dataset with 8 random clusters
-    X, y = make_blobs(centers=8, n_features=12, shuffle=True, random_state=42)
+    # Generate synthetic dataset with 8 random clusters
+    X, y = make_blobs(n_samples=1000, n_features=12, centers=8, random_state=42)
```
I removed `shuffle=True` since it is `True` by default.

In addition, I added `n_samples=1000` since that is what it was set to in the original script used to generate the image, and if not included, `n_samples` is only equal to `100` by default.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -8,21 +8,23 @@ The ``KElbowVisualizer`` implements the "elbow" method to help data scientists s
 
-    visualizer.fit(X)    # Fit the data to the visualizer
-    visualizer.poof()    # Draw/show/poof the data
+    visualizer.fit(X)        # Fit the data to the visualizer
```
I like to add 8 spaces between the code and inline comments in the docs, since I think it makes it easier to read.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -31,21 +33,25 @@ However, two other metrics can also be used with the ``KElbowVisualizer`` -- ``s
 The ``KElbowVisualizer`` also displays the amount of time to train the clustering model per :math:`K` as a dashed green line, but is can be hidden by setting ``timings=False``. In the following example, we'll use the ``calinski_harabaz`` score and hide the time to fit the model.
 
 .. plot::
+    :context: close-figs
```
Added new plot directive settings.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -31,21 +33,25 @@ However, two other metrics can also be used with the ``KElbowVisualizer`` -- ``s
-    # Create synthetic dataset with 8 random clusters
-    X, _ = make_blobs(centers=8, n_features=12, shuffle=True, random_state=42)
+    # Generate synthetic dataset with 8 random clusters
+    X, y = make_blobs(n_samples=1000, n_features=12, centers=8, random_state=42)
```
I replaced `_` with `y` here, since even though the target variable is not needed and it doesn't affect the final plot, I thought `_` could be confusing for new users who may not have seen this before. In addition, it's not used in the other examples in our documentation when `make_blobs` is used to create the dataset.

I also removed `shuffle=True` and added `n_samples=1000` for the same reasons I described above in a previous comment.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -31,21 +33,25 @@ However, two other metrics can also be used with the ``KElbowVisualizer`` -- ``s
 
-    visualizer.fit(X)    # Fit the data to the visualizer
-    visualizer.poof()    # Draw/show/poof the data
+    visualizer.fit(X)        # Fit the data to the visualizer
```
Added additional spacing.

Kautumn06 commented on docs/api/cluster/elbow.rst:
```diff
@@ -31,21 +33,25 @@ However, two other metrics can also be used with the ``KElbowVisualizer`` -- ``s
     # Instantiate the clustering model and visualizer
     model = KMeans()
-    visualizer = KElbowVisualizer(model, k=(4,12), metric='calinski_harabaz', timings=False)
+    visualizer = KElbowVisualizer(
```
Small formatting change since the line was quite long.

Kautumn06 commented on docs/api/cluster/icdm.rst:
```diff
@@ -5,26 +5,24 @@ Intercluster Distance Maps
 Intercluster distance maps display an embedding of the cluster centers in 2 dimensions with the distance to other centers preserved. E.g. the closer to centers are in the visualization, the closer they are in the original feature space. The clusters are sized according to a scoring metric. By default, they are sized by membership, e.g. the number of instances that belong to each center. This give ...
 
-.. code:: python
+.. plot::
```
Previously, this example in the documentation had been broken out into two blocks&mdash;one to generate the dataset and one to `fit` the model and `poof` the visualizer. So in order to have it work with the new plot directive, I combined them into one and moved all of the import statements to the top of the block.

Kautumn06 commented on docs/api/cluster/icdm.rst:
```diff
@@ -5,26 +5,24 @@ Intercluster Distance Maps
     from yellowbrick.cluster import InterclusterDistance
 
-    # Instantiate the clustering model and visualizer
-    visualizer = InterclusterDistance(KMeans(9))
```
As you'll see in the next comment, I broke this code out into two lines to keep it consistent with the rest of the examples in the documentation.

Kautumn06 commented on docs/api/cluster/icdm.rst:
```diff
@@ -5,26 +5,24 @@ Intercluster Distance Maps
+    # Generate synthetic dataset with 12 random clusters
+    X, y = make_blobs(n_samples=1000, n_features=12, centers=12, random_state=42)
 
+    # Instantiate the clustering model and visualizer
```
This code had previously been combined in a single line, and while that does work, I think it's clearer to have it in two. In addition, this is the format we could throughout the rest of the documentation so I wanted to be consistent.

Kautumn06 commented on docs/api/cluster/icdm.rst:
```diff
@@ -5,26 +5,24 @@ Intercluster Distance Maps
-    visualizer.fit(X) # Fit the training data to the visualizer
-    visualizer.poof() # Draw/show/poof the data
+    # Generate synthetic dataset with 12 random clusters
+    X, y = make_blobs(n_samples=1000, n_features=12, centers=12, random_state=42)
```
Originally, the `.rst` file showed `n_features=16`; however, in the script used to generate the image, the parameter was set to `n_features=12`, so I changed it here to reflect that.

Kautumn06 commented on docs/api/cluster/silhouette.rst:
```diff
@@ -7,29 +7,23 @@ The Silhouette Coefficient is used when the ground-truth about the dataset is un
     from yellowbrick.cluster import SilhouetteVisualizer
 
+    # Generate synthetic dataset with 8 random clusters
+    X, y = make_blobs(n_samples=1000, n_features=12, centers=8, random_state=42)
```
Added the parameters used in the original script to generate the image.

Kautumn06 commented on docs/api/cluster/silhouette.rst:
```diff
@@ -7,29 +7,23 @@ The Silhouette Coefficient is used when the ground-truth about the dataset is un
+
     # Instantiate the clustering model and visualizer
-    model = MiniBatchKMeans(6)
+    model = KMeans(6)
```
Using `KMeans` instead of `MiniBatchKMeans` didn't affect the plot so I changed it here to make it consistent with our other clustering visualizers. In addition, I was worried that new users might be confused about why it is used here and not in the other clustering examples.

rebeccabilbro commented:
Hi @Kautumn06 &mdash; sorry that I didn't get to this earlier today, but we've been sorting out some problems with our automated tests. Now that everything seems to be fixed, I've updated your branch so that the tests can run again; I'll take a look at your updates tomorrow!

Kautumn06 commented:
Hi @rebeccabilbro no problem! I haven't pushed any changes since yesterday afternoon so hopefully there shouldn't be anymore problems. However, just let me know if you have any questions or if there is anything else you need me to help with!

rebeccabilbro commented:
Hi @Kautumn06 - I'm starting to take a look at your updates to the clustering docs and I had a quick question &mdash; in your above message, you said:

> when I ran `make html`, the tests failed.

and I just wanted to make sure I understood what behavior you were experiencing before proceeding. 

When you run `make html`, it builds a copy of the docs locally on your machine, but it doesn't run any tests. However, oftentimes when we run the `make html` command after making experimental changes to the docs, we'll get warnings or error messages from Sphinyx in the command line that tell us that there's something funky going on in the docs. When you say the tests failed after you ran `make html`, do you mean you got Sphinyx warnings? This is probably a sign that there is some kind of formatting error in the rst, a missing reference link, etc.

Alternatively, if you mean that after making the updates to the docs and running the tests (either locally on your machine with `pytest`, or by opening the PR, which automatically runs the tests and shows the results here on Github), you observed failing tests on the command line or saw the little red X mark here on GH, that's possibly due to the [miniconda/Appveyor problems we were experiencing](https://github.com/DistrictDataLabs/yellowbrick/pull/743) with our continuous integration tools). 

The continuous integration testing issue is now resolved (for now ;D ), so would you please let me know if what you experienced seemed instead like a Sphinx issue, so I can look into that during my code review? Thanks for working on this!

Kautumn06 commented:
Hi @rebeccabilbro sorry for any confusion! The test I was referring to was in fact the red X here on GitHub. Originally, I had thought that the problem occurred because of the parameter issue I mentioned above; however, it actually turned out to be from the [miniconda/Appveyor problem](https://github.com/DistrictDataLabs/yellowbrick/pull/743) we've been experiencing.

So please just let me know if that answers your question or if there is anything else you need me to do!

Kautumn06 commented:
Hi @rebeccabilbro  I just wanted to let you know that I fixed the two minor merge conflicts that came up today after #729 was merged in. It was a quick fix since the two differences were from me including the new plot directives `:alt:` option in the code blocks. Let me know if you have any questions!

Kautumn06 commented:
Hi @rebeccabilbro I've replaced `KMeans` with `MiniBatchKMeans` in the `SilhouetteVisualizer` documentation example. I also added the plot directive `:alt:` option to the code block since I had originally forgot to add it. Once the tests have passed, I'll merge it in.
